File: Grind 75 in Java/0207-course-schedule.java
class Solution {
    public boolean canFinish(int n, int[][] pre) {
        ArrayList<Integer>[] adj = new ArrayList[n];

        for (int i = 0; i < n; i++)
            adj[i] = new ArrayList<>();

        for (int[] p : pre)
            adj[p[0]].add(p[1]);

        boolean[] visit = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!dfs(i, visit, adj)) return false;
        }
       
        return true;
    }

    private boolean dfs(int i, boolean[] visit, ArrayList<Integer>[] adj) {
        if (visit[i]) return false;
        if (adj[i].isEmpty()) return true;

        visit[i] = true;

        for (int next : adj[i]) {
            if (!dfs(next, visit, adj)) return false;
        }

        visit[i] = false;
        adj[i].clear();
        return true;
    }
}

File: leetcode in kotlin/0207-course-schedule.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val edges = Array<ArrayList<Int>> (numCourses) { ArrayList() }
        val visit = HashSet<Int>()

        prerequisites.forEach {
            val (a,b) = it
            edges[a].add(b)
        }

        fun dfs(i: Int): Boolean {
            if (i in visit) return false
            if (edges[i].size == 0) return true

            visit.add(i)
            edges[i].forEach {
                if (!dfs(it)) return false
            }

            visit.remove(i)
            edges[i].clear()
            return true
        }

        for (i in 0..edges.size-1) {
            if (!dfs(i)) return false
        } 

        return true
    }
}
